rvzr/model_dynamorio/backend/speculator_abc.cpp

```diff
@@ -108,9 +108,35 @@ pc_t SpeculatorABC::rollback(dr_mcontext_t *mc)
         // Try restoring the previous value in memory.
         size_t w_size = 0;
         bool success = dr_safe_write((byte *)store.addr, store.size, &store.val, &w_size);
+
+        // If the rollback did not succeed, check the reason.
+        //
+        // NOTE: Rollbacks can fail in the following cases:
+        //       1. Rollback of an invalid store: the store should have caused an exception,
+        //       therefore the corresponding entry should not be marked as committed. We
+        //       just ignore these cases, as they are expected during speculation.
+        //       2. Rollback after faulty indirect call/ret: if a data ddress is used as an
+        //       (invalid) call or ret target, DynamoRIO will wrongly mark the address as belonging
+        //       to the code cache, and therefore prohibit writes.
+        if (not success and not(store_log.size() > last_committed)) {
```

Comment:
This starts to resemble the Hadouken Pattern. Perhaps move this into a separate private method?